### PR TITLE
add: schemaCompiler property

### PR DESCRIPTION
**Pull request by Eomm**
Hi,
I would like to add a `getSchemaCompiler` method in order to get returned the function used by the `validation.js` to build the validation function.
That `this._schemaCompiler` have all the JSON schemas of the fastify instance and could let the devs to add some feature validation using the schemas they already set plus the resolved shared-schema.

I'm adding this method in order to solve issue #1376 
I have already build a plugin that accomplishes that task but it works only thanks to this line:

https://github.com/Eomm/fastify-schema-constraint/blob/dbec55b3da4dc88d8c4a1a7c1d337396cc4e9757/plugin.js#L84

I think that could give some new flexibility to the community 👍

Let me know what you think about this PR.
I'm not sure if there could be some side effect I don't spot

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Eomm** [comment]
> Shouldn't be easier just expose it in the public API?

Sure! But it could be changed at runtime that could have side effects (it is already possible because it is "private" 😁).
Maybe `_schemaCompiler ` isn't public only because we want to communicate to devs that `_schemaCompiler` can't be manipulated freely.

File: test/post.test.js [hunk 1]
```diff
@@ -23,3 +23,15 @@ t.test('cannot set schemaCompiler after binding', t => {
...
   })
 })
+
+t.test('get schemaCompiler after set schemaCompiler', t => {
+  t.plan(2)
+  const mySchemaCompiler = () => { }
+
+  const fastify = Fastify()
+  fastify.setSchemaCompiler(mySchemaCompiler)
+  const sc = fastify.getSchemaCompiler()
```
Comment (mcollina) [inline test/post.test.js]: I think we should refactor these to be getter and setters for the schemaCompiler property.

Comment (Eomm) [reply test/post.test.js]: Do you mean to add the setter removing the `setSchemaCompiler`?
Wouldn't this be a breaking change?

**Comment by mcollina** [comment]
You can keep them both

Il giorno gio 14 mar 2019 alle 20:27 Manuel Spigolon <
notifications@github.com> ha scritto:

File: fastify.js [hunk 2]
```diff
@@ -184,6 +184,15 @@ function build (options) {
     setErrorHandler: setErrorHandler
   }
 
+  Object.defineProperty(fastify, 'schemaCompiler', {
+    get: function () {
+      return this._schemaCompiler
```
Comment (mcollina) [inline fastify.js]: I would put _schemaCompiler behind a symbol.

File: docs/Server.md [hunk 3]
```diff
@@ -491,6 +491,10 @@ To learn more, see [shared schema example](https://github.com/fastify/fastify/bl
 #### setSchemaCompiler
 Set the schema compiler for all routes [here](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#schema-compiler).
 
+<a name="schema-compiler"></a>
+#### schemaCompiler
+Get the schema compiler for all routes if one has been settled.
```
Comment (mcollina) [inline docs/Server.md]: you should also document setting the schemaCompiler via the property.

**Comment by Eomm** [comment]
@delvedor, are changes good for your initial comment?

File: docs/Server.md [hunk 4]
```diff
@@ -526,6 +526,11 @@ To learn more, see [shared schema example](https://github.com/fastify/fastify/bl
 #### setSchemaCompiler
 Set the schema compiler for all routes [here](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#schema-compiler).
 
+<a name="schema-compiler"></a>
+#### schemaCompiler
+This property can be used to set the schema compiler, it is a shortcut for the `setSchemaCompiler` method,
+and get the schema compiler back for all routes.
```
Comment (delvedor) [inline docs/Server.md]: Can you remove the line break here?

**Pull request merged**